File: scripts/changeLogHelper.py
import re
from typing import List


class ChangeLogHelper:
    def __init__(self,
                 version_regex: str,
                 date_regex: str,
                 cli_core_changelog_filename: str = 'CHANGES.md',
                 oai_changelog_filename: str = 'OAI-CHANGES.md'):
        self.version_regex = re.compile(version_regex)
        self.date_regex = re.compile(date_regex)
        self.cli_core_changelog_filename = cli_core_changelog_filename
        self.oai_changelog_filename = oai_changelog_filename

    def get_first_two_versions(self) -> List[str]:  # oai_changes.md
        versions = []
        for line in self.get_changelog_lines(self.oai_changelog_filename):
            found = self.version_regex.search(line)
            if found:
                versions.append(found.group())
                if len(versions) == 2:
                    break
        if len(versions) == 0:
            print('There are no versions detected in the given changelog file:' + self.oai_changelog_filename)
        else:
            print("Versions:")
            print(*versions)
        return versions

    def get_latest_changelog_generated_date(self) -> str:
        latest_date = None

        for line in self.get_changelog_lines(self.cli_core_changelog_filename):
            found = self.date_regex.search(line)
            if found:
                latest_date = found.group()
                break

        if latest_date:
            print(f'Detected latest date: {latest_date}')
        else:
            print('Did not detect any dates')

        return latest_date

    def get_changes_after_given_date(self, date: str) -> str:
        read_lines = False
        file_data = ''
        for line in self.get_changelog_lines(self.oai_changelog_filename):
            found = self.date_regex.search(line)
            if found:
                current_date = found.group()
                if current_date > date:
                    read_lines = True
                else:
                    break
            elif read_lines:
                file_data += line
        return file_data

    def append_changes_to_changelog(self):
        latest_date = self.get_latest_changelog_generated_date()  # changes.md
        if latest_date:
            changeLog = self.get_changes_after_given_date(latest_date)  # oai_changes.md
            if changeLog:
                with open(self.cli_core_changelog_filename, 'r+') as f:  # changes.md
                    f.readline()
                    f.readline()
                    pos = f.tell()
                    content = f.read()
                    f.seek(pos, 0)
                    f.write(changeLog.rstrip('\r\n') + '\n' + content)
            else:
                print("No changes to write")
        else:
            print("No latest date found")

    def get_changelog_lines(self, changelog_filename: str) -> List[str]:
        with open(changelog_filename) as changelog:
            return changelog.readlines()

File: scripts/change-log-helper.js
const fs = require('fs');
const readline = require('readline');

const { logger } = require('../src/services/messaging/logging');

const defaultVersionRegex = /(\d+)\.(\d+)\.(\d+)/;
const defaultDateRegex = /\d{4}\-(0[1-9]|1[012])\-(0[1-9]|[12][0-9]|3[01])/;

class ChangeLogHelper {
  constructor(
    cliCoreChangelogFilename,
    oaiChangelogFilename,
    versionRegex = defaultVersionRegex,
    dateRegex = defaultDateRegex,
  ) {
    this.versionRegex = versionRegex;
    this.dateRegex = dateRegex;
    this.cliCoreChangelogFilename = cliCoreChangelogFilename;
    this.oaiChangelogFilename = oaiChangelogFilename;
    this.logger = logger;
  }

  async getAllReleaseVersionsFromGivenDate(date) {
    this.logger.debug(`Started detecting the versions from the date: ${date}`);
    const versions = [];
    const readLine = await this.getReadLiner(this.oaiChangelogFilename);
    for await (const line of readLine) {
      const currentDate = this.dateRegex.exec(line);
      if (currentDate) {
        const version = this.versionRegex.exec(line);
        if (version) {
          versions.push(version[0]);
        }
        if (currentDate[0] <= date) {
          break;
        }
      }
    }
    this.logger.debug(`Detected Versions: ${versions}`);
    return versions;
  }

  async getLatestChangelogGeneratedDate() {
    this.logger.debug('Started detecting the latest date in cli core changelog');
    let latestDate;
    const readLine = await this.getReadLiner(this.cliCoreChangelogFilename);
    for await (const line of readLine) {
      latestDate = this.dateRegex.exec(line);
      if (latestDate) {
        latestDate = latestDate[0];
        this.logger.debug(`Detected the latest Date: ${latestDate}`);
        break;
      }
    }
    return latestDate;
  }

  async getChangesAfterGivenDate(date) {
    this.logger.debug(`Started getting the changelog from given date: ${date}`);
    let readLines = false;
    let fileData = '';
    const readLine = await this.getReadLiner(this.oaiChangelogFilename);
    for await (const line of readLine) {
      const currentDate = this.dateRegex.exec(line);
      if (currentDate) {
        if (currentDate[0] > date) {
          this.logger.debug('Reading the lines');
          readLines = true;
        } else {
          this.logger.debug(`Changes from OpenAPI specs: ${fileData}`);
          break;
        }
      } else if (readLines) {
        fileData += `${line}\n`;
      }
    }
    return fileData;
  }

  async getAndAppendChangesToChangelog() {
    this.logger.debug('Started getAndAppendChangesToChangelog');
    const latestDate = await this.getLatestChangelogGeneratedDate(); // changes.md
    if (latestDate) {
      const changeLog = await this.getChangesAfterGivenDate(latestDate); // oai_changes.md
      if (changeLog) {
        try {
          this.logger.debug('Updating the CHANGES.md');
          const data = fs.readFileSync(this.cliCoreChangelogFilename);
          const fd = fs.openSync(this.cliCoreChangelogFilename, 'w+');
          const insert = Buffer.from(changeLog);
          fs.writeSync(fd, insert, 0, insert.length, 0);
          fs.writeSync(fd, data, 0, data.length, insert.length);
          fs.close(fd, (err) => {
            if (err) throw err;
          });
        } catch (error) {
          this.logger.error(`Error while updating the changelog: ${error}`);
        }
        return changeLog;
      }
    }
    return '';
  }

  async getReadLiner(filename) {
    const fileStream = fs.createReadStream(filename);
    return readline.createInterface({
      input: fileStream,
    });
  }
}
module.exports = {
  ChangeLogHelper,
};
